crates/holochain/src/conductor/conductor.rs

```diff
@@ -742,53 +758,88 @@ impl Conductor {
-                    } else {
-                        Err(ConductorError::AppNotRunning(app_id.clone()))
+                    let app = state.get_app_mut(&app_id)?;
+                    let app_role_assignment = app
+                        .roles()
+                        .get(&role_id)
+                        .ok_or_else(|| AppError::AppRoleIdMissing(role_id.to_owned()))?;
+                    if app_role_assignment.is_clone_limit_reached() {
+                        return Err(ConductorError::AppError(AppError::CloneLimitExceeded(
+                            app_role_assignment.clone_limit(),
+                            app_role_assignment.clone(),
+                        )));
                     }
+                    let parent_dna_hash = app_role_assignment.dna_hash().clone();
+                    Ok((state, parent_dna_hash))
                 }
             })
             .await?;
+        // clone cell from base cell DNA
         let child_dna = ribosome_store.share_ref(|ds| {
-            ds.get_dna_file(&parent_dna_hash)
-                .ok_or(DnaError::DnaMissing(parent_dna_hash))?
-                .modify_phenotype(random_network_seed(), properties)
+            let mut dna_file = ds
+                .get_dna_file(&base_cell_dna_hash)
+                .ok_or(DnaError::DnaMissing(base_cell_dna_hash))?
+                .modify_phenotype(dna_phenotype);
+            if let Some(name) = name {
+                dna_file = dna_file.set_name(name);
+            }
+            Ok::<_, DnaError>(dna_file)
         })?;
         let child_dna_hash = child_dna.dna_hash().to_owned();
         let child_ribosome = RealRibosome::new(child_dna)?;
         self.register_phenotype(child_ribosome);
-        let (_, cell_id) = self
+        // add clone cell to app and increment clone index
+        let (_, installed_clone_cell) = self
             .update_state_prime(move |mut state| {
-                if let Some(app) = state.installed_apps_mut().get_mut(&app_id) {
-                    let agent_key = app.role(&role_id)?.agent_key().to_owned();
-                    let cell_id = CellId::new(child_dna_hash, agent_key);
-                    app.add_clone(&role_id, cell_id.clone())?;
-                    Ok((state, cell_id))
-                } else {
-                    Err(ConductorError::AppNotRunning(app_id.clone()))
+                let app = state.get_app_mut(&app_id)?;
+                let agent_key = app.role(&role_id)?.agent_key().to_owned();
+                let cell_id = CellId::new(child_dna_hash, agent_key);
+                let next_clone_index = app.next_clone_index(&role_id)?;
+                let clone_id = CloneId::new(&role_id, next_clone_index);
+                app.add_clone(&clone_id, &cell_id)?;
+                let installed_clone_cell =
+                    InstalledCell::new(cell_id, clone_id.as_app_role_id().clone());
+                app.increment_next_clone_index(&role_id)?;
```

Comment:
Adding the clone should happen in the same state update as checking if the clone limit is exceeded, or else we invite a race condition. Maybe there is a performance reason to break it up, but it seems more important to maintain atomicity.

Reply:
It was also due to performance to move this check to the calling function, but more importantly because otherwise a clone cell DNA is created and I'm not sure what'll happen with it if it's not then connected to an app. There's no reconciliation step that'd clean it up. What do you think?

Reply:
I think it's ok @jost-s 

Performance-wise, registering the dna is a small lift. And, I don't think there's any logical harm in having an unused dna that's been registered, other than the memory it takes up. There should be a cleanup step for that in case of error though, to be tidy.

I think that's better than inviting the race condition, especially considering that someone might reasonably want to create many clone cells at once, which would be exactly when we would expect the race condition to occur. That might be an argument for a batch creation call, but it seems premature to add it now.

Reply:
I cleaned up the conductor state update and only kept one check for clone limit. It feels really clean now. Please review again @maackle